test/functional/bind.cpp

```diff
@@ -0,0 +1,387 @@
+                     void>::value);
+  bound_void(std::forward<Args>(args)...);
+}
+
+template <typename R,
+          typename Bound,
+          typename BoundR,
+          typename BoundVoid,
+          typename... Args>
+void check_bound_const(R&& r,
+                       Bound const bound,
+                       BoundR const bound_r,
+                       BoundVoid const bound_void,
+                       Args&&... args) {
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound)>>::value);
+  CHECK(bound(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_r)>>::value);
+  CHECK(std::is_same<decltype(bound_r(std::forward<Args>(args)...)), R>::value);
+  CHECK(bound_r(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_void)>>::value);
+  CHECK(std::is_same<decltype(bound_void(std::forward<Args>(args)...)),
+                     void>::value);
+  bound_void(std::forward<Args>(args)...);
+}
+
+template <typename R, typename F, typename Arg1>
+void check_bind(R&& r, F&& f, Arg1&& arg1) {
+  namespace ph = slb::placeholders;
+  return check_bound<R>(std::forward<R>(r),
+                        slb::bind(std::forward<F>(f), ph::_1),
+                        slb::bind<R>(std::forward<F>(f), ph::_1),
+                        slb::bind<void>(std::forward<F>(f), ph::_1),
+                        std::forward<Arg1>(arg1));
+}
+
+template <typename R, typename F, typename Arg1>
+void check_bind_const(R&& r, F&& f, Arg1&& arg1) {
+  namespace ph = slb::placeholders;
+  return check_bound_const<R>(std::forward<R>(r),
+                              slb::bind(std::forward<F>(f), ph::_1),
+                              slb::bind<R>(std::forward<F>(f), ph::_1),
+                              slb::bind<void>(std::forward<F>(f), ph::_1),
+                              std::forward<Arg1>(arg1));
+}
+
+template <typename R, typename F, typename Arg1, typename Arg2>
+void check_bind(R&& r, F&& f, Arg1&& arg1, Arg2&& arg2) {
+  namespace ph = slb::placeholders;
+  return check_bound<R>(std::forward<R>(r),
+                        slb::bind(std::forward<F>(f), ph::_1, ph::_2),
+                        slb::bind<R>(std::forward<F>(f), ph::_1, ph::_2),
+                        slb::bind<void>(std::forward<F>(f), ph::_1, ph::_2),
+                        std::forward<Arg1>(arg1),
+                        std::forward<Arg2>(arg2));
+}
+
+TEST_CASE("bind(mem-obj-ptr)", "[func.bind.bind]") {
+  auto f = &C::obj;
+
+  /* forwarding call wrapper */ {
+    using call_wrapper = decltype(f);
+
+    CHECK(std::is_move_constructible<call_wrapper>::value);
+    CHECK(std::is_copy_constructible<call_wrapper>::value);
+  }
+
+  /* reference */ {
+    C x = {42};
+    C& r = x;
+    C const& cr = x;
+
+    CHECK_NESTED(bind<int&>(r.obj, f, r));
```

Comment:
Maybe we should add some more `bind_const` tests?

```
  CHECK_NESTED(bind_const<const int&>(r.obj, f, r));
```
For example is interesting since the `r` gets decay-copied into the bound object
which changes behavior when the bound object itself is `const`.

Reply:
This is still pending...